File: code/src/leetcode/RangeFrequencyQueries.java
package leetcode;
import java.util.HashMap;

public class RangeFrequencyQueries {
    private static class RangeFreqQuery {

        private static class SegmentTreeNode {
            int start, end; // Range represented by this node
            HashMap<Integer, Integer> freqAcc;      // Value stored in this node (e.g., sum of the range)
            SegmentTreeNode left, right; // Left and right child nodes

            // Constructor for a node
            public SegmentTreeNode(int start, int end) {
                this.start = start;
                this.end = end;
                this.freqAcc = new HashMap<Integer, Integer>(); // Initialize the value to 0
                this.left = null;
                this.right = null;
            }
        }
        private SegmentTreeNode root;

        // Recursive function to build the segment tree
        private SegmentTreeNode buildTree(int[] arr, int start, int end) {
            if (start > end) {
                return null; // Base case: invalid range
            }

            // Create a new node for the current range
            SegmentTreeNode node = new SegmentTreeNode(start, end);

            if (start == end) {
                // Leaf node: stores the value of a single element
                node.freqAcc = new HashMap<>();
                node.freqAcc.put(arr[start], 1);
                return node;
            }

            int mid = start + (end - start) / 2;

            // Recursively build the left and right subtrees
            node.left = buildTree(arr, start, mid);
            node.right = buildTree(arr, mid + 1, end);

            // Internal node: stores the sum of its children
            node.freqAcc = new HashMap<>(node.left.freqAcc);
            // The right node might not exist (can be null). Handle null.
            if (node.right != null) {
                for (var k: node.right.freqAcc.keySet()) {
                    node.freqAcc.merge(k, node.right.freqAcc.get(k), Integer::sum);
                }
            }

            return node;
        }

        public RangeFreqQuery(int[] arr) {
            this.root = buildTree(arr, 0, arr.length - 1);
        }

        // Recursive function to query the segment tree
        private int queryTree(SegmentTreeNode node, int queryLeft, int queryRight, int value) {
            if (node == null || queryRight < node.start || queryLeft > node.end) {
                // Fully outside the query range
                return 0;
            }

            if (queryLeft <= node.start && node.end <= queryRight) {
                // Fully inside the query range
                return node.freqAcc.getOrDefault(value, 0);
            }

            // Partially overlapping: split and recurse
            return queryTree(node.left, queryLeft, queryRight, value) +
                    queryTree(node.right, queryLeft, queryRight, value);
        }

        public int query(int left, int right, int value) {
            return queryTree(root, left, right, value);
        }
    }
}

File: code/src/leetcode/RangeFrequencyQueries.py
from collections import Counter
from dataclasses import dataclass
from typing import List, Optional


@dataclass
class SegmentTreeNode:
    start: int
    end: int
    freq: Counter[int]
    left: Optional['SegmentTreeNode'] = None
    right: Optional['SegmentTreeNode'] = None


class RangeFreqQuery:
    def __init__(self, arr: List[int]):
        self.root = self._build_tree(arr, 0, len(arr) - 1)
    
    def _build_tree(self, arr: List[int], start: int, end: int) -> Optional[SegmentTreeNode]:
        if start > end:
            return None
            
        node = SegmentTreeNode(start=start, end=end, freq=Counter())
        
        if start == end:
            node.freq[arr[start]] = 1
            return node
            
        mid = (start + end) // 2
        node.left = self._build_tree(arr, start, mid)
        node.right = self._build_tree(arr, mid + 1, end)
        
        # Merge frequencies from children
        node.freq = Counter()
        if node.left:
            node.freq.update(node.left.freq)
        if node.right:
            node.freq.update(node.right.freq)
            
        return node
    
    def query(self, left: int, right: int, value: int) -> int:
        def _query(node: Optional[SegmentTreeNode], query_left: int, query_right: int) -> int:
            if not node or query_right < node.start or query_left > node.end:
                return 0
                
            if query_left <= node.start and node.end <= query_right:
                return node.freq[value]
                
            return (_query(node.left, query_left, query_right) + 
                   _query(node.right, query_left, query_right))
        
        return _query(self.root, left, right)
